plugins/node/opentelemetry-instrumentation-pino/README.md

```diff
@@ -15,60 +15,105 @@ Compatible with OpenTelemetry JS API and SDK `1.0+`.
+});
 ```
 
-### Fields added to pino log objects
+### Log sending
+
+Creation of a Pino Logger will automatically add a [Pino destination](https://getpino.io/#/docs/api?id=pinooptions-destination-gt-logger) that sends log records to the OpenTelemetry Logs SDK. The OpenTelemetry SDK can be configured to handle those records -- for example, sending them on to an OpenTelemetry collector for log archiving and processing. The example above shows a minimal configuration that emits OpenTelemetry log records to the console for debugging.
+
+If the OpenTelemetry SDK is not configured with a Logger provider, then this added destination will be a no-op.
+
+Log sending can be disabled with the `disableLogSending: true` option.
 
-For the current active span, the following fields are injected. These field names can be optionally configured via `logKeys` in the PinoInstrumentation config:
+### Log correlation
+
+Pino logger calls in the context of a tracing span will have fields identifying the span added to the log record. This allows [correlating](https://opentelemetry.io/docs/specs/otel/logs/#log-correlation) log records with tracing data. The added fields are ([spec](https://opentelemetry.io/docs/specs/otel/compatibility/logging_trace_context/)):
 
 - `trace_id`
 - `span_id`
 - `trace_flags`
 
+These field names can optionally be configured via the `logKeys` option. For example:
+
+```js
+    new PinoInstrumentation({
+      logKeys: {
+        traceId: 'myTraceId',
+        spanId: 'mySpanId',
+        traceFlags: 'myTraceFlags',
+      },
+    }),
+```
+
+After adding these fields, the optional `logHook` is called to allow injecting additional fields. For example:
+
+```js
+  logHook: (span, record) => {
+    record['resource.service.name'] = provider.resource.attributes['service.name'];
+  }
+```
+
 When no span context is active or the span context is invalid, injection is skipped.
+Log injection can be disabled with the `disableLogCorrelation: true` option.
 
-### Supported versions
+### Pino instrumentation options
 
-`>=5.14.0 <10`
+| Option                  | Type              | Description |
+| ----------------------- | ----------------- | ----------- |
+| `disableLogSending`     | `boolean`         | Whether to disable [log sending](#log-sending). Default `false`. |
+| `disableLogCorrelation` | `boolean`         | Whether to disable [log correlation](#log-correlation). Default `false`. |
+| `logKeys`               | record            | A record with keys `traceId`, `spanId`, and `traceFlags` string fields giving the field names to use for log-correlation span data. |
+| `logHook`               | `LogHookFunction` | An option hook to inject additional context to a log record after trace-context has been added. This requires `disableLogCorrelation` to be false. |
 
 ## Semantic Conventions
 
 This package does not currently generate any attributes from semantic conventions.
 
+## Alternative log sending with `pino-opentelemetry-transport`
```

Comment:
I wonder if the instrumentation would be capable to know if `pino-opentelemetry-transport` is used when creating a new logger and disable sending in the main thread. Or maybe warn the user if this transport is detected and suggest to change its `disableSending` configuration if applies.

Reply:
First, I don't believe there is a way for instrumentation-pino to know if `pino-opentelemetry-transport` is being used.  Here is an example script:

```
const otel = require('@opentelemetry/api');
const pino = require('pino');

const transport = pino.transport({ target: 'pino-opentelemetry-transport' });
const logger = pino({level: 'debug'}, transport);
console.log('XXX The logger stream is:', logger[pino.symbols.streamSym]);

logger.debug({foo: 'bar'}, 'hi');

const tracer = otel.trace.getTracer('example');
tracer.startActiveSpan('manual-span', span => {
  logger.info('in a span');
  span.end();
});
```

If `instrumentation-pino` is active, the only thing it has to look at is that `logger[pino.symbols.streamSym]` value -- the destination stream of the Pino Logger. This is the thing that the instrumentation calls `origStream` and wraps in a `multistream`.

That thing is a `ThreadStream` (from https://github.com/pinojs/thread-stream/blob/main/index.js):

```
XXX The logger stream is: <ref *1> ThreadStream {
  _events: [Object: null prototype] {
    message: [Function (anonymous)],
    ready: [Function: onReady],
    close: [Function (anonymous)]
  },
  _eventsCount: 3,
  _maxListeners: undefined,
  worker: Worker {
    _events: [Object: null prototype] {
      newListener: [Function (anonymous)],
      removeListener: [Function (anonymous)],
      message: [Function: onWorkerMessage],
      exit: [Function: onWorkerExit]
    },
    _eventsCount: 4,
    _maxListeners: undefined,
    performance: { eventLoopUtilization: [Function: bound eventLoopUtilization] },
    stream: FakeWeakRef { _value: [Circular *1] },
    [Symbol(kCapture)]: false,
...
```

That `ThreadStream` provides no way to know what the `target` package or Node.js file is being run in the Worker thread.  In other words, instr-pino can know that a Pino *transport* is being used (i.e. something running in a separate thread), but has no idea whether that is `pino-opentelemetry-transport`.

I think it would be overkill to look into instrumenting `pino.transport()` to tag any returned transport that is using `pino-opentelemetry-transport` to try to tell.

---

Second, I'm not sure that we'd want instr-pino to automatically do anything even if it *did* know.

- Automatically disabling log-sending if `pino-opentelemetry-transport` (POT) was detected would potentially be surprising behaviour. I think to be explicit we'd then want `disableLogSending` to be a tri-state config: true, false, "true-if-POT-detected".
- Automatically warning if POT was detected? I suppose this would be fair. *Most* users would never want both active on a given Pino logger. (Eventually perhaps we'd want the ability for users to silence particular warnings from the OTel SDK that they don't care above -- something like Python's warnings filter https://docs.python.org/3/library/warnings.html#the-warnings-filter)

These are why I opted for just have documentation about this so that users are at least made aware of the options (assuming they get as far as reading this README). Perhaps a future "Troubleshooting" guide would merit a section on  "I'm using Pino and getting double logs. Why?!".

Reply:
Those are valid points. Thanks for taking the time to give your reasoning :)